### ISSUE TITLE
Add mod configuration options

### ISSUE THREAD

**Issue by Tahvohck** [Spacehaven-modding-tools/spacehaven-modloader#4]
Requested by Killparadise in the Discord server. User-side configuration of the mod, separate from the mod data files.

> I'd say just letting the user tweak the config for instance on my lower system points mod, I could set the defaults to 1 and they can config it to whatever they want, no need to upload multiple versions of the same thing

Thoughts:
- Start with just numerical config
- Allow config to specify min/max/neither
- Store data in the ModDatabase entry for the mod
- Support patch mods only? Supporting merge mods would require parsing xml for everything, while patch mods can use an attribute on the <value> tag

**Comment by WalterLPearson** [comment]
### OVERVIEW
There are the Mod Developer (Dev) and the Mod User/Player (User).

If we make an XML Attribute that flags data as modifiable by specifying  data type and (optional) valid range / regex check, then this can be parsed easily at Mod Load time.
It will be the responsibility of the Dev to mark which fields can be changed, or need to be configured.
The User will simply enter valid values on the GUI.

Once that first step is done, implementing more complexity (e.g. give the user a checkbox that impacts many different changes) can be added to the code.  Using named variables and applying basic math with them comes to mind.

We are in danger of trying to implement full scripting, be it by running Python or implementing ECMA script support.
**Both scripting options should be avoided.**

Clarification: WP is in danger of getting carried away and trying to implement script in the Mod Loader.

**Comment by WalterLPearson** [comment]
### CODE DESIGN PROPOSALS

**Option A:**
The patch values themselves have annotations specifying data type (e.g. int,float,string,bool) and validation (e.g. number range, string regex) and a label seen by the user.  These can be placed on any XML element, in both the merge and patch stages, permitting any detail to be modified in any way.
This requires mods to be fully parsed at load time, instead of at merge time.

**Option B:**
The info.xml file has all configuration.  The Dev is responsible for specifying config options, giving them a unique variable name along with description (for User), data type, and validation.
This variable is then used at Patch time and replaced when found in "value".

**Option C:**
A third option may be simply moving directly to implementing merge-and-patch-time scripting.  Which should be avoided.

I suggest Option B, because it is the most incremental change while remaining extendable, and can definitely be completed for v0.10.0

**Comment by Tahvohck** [comment]
Go for option B. The info file is the only file that's currently read on every launch, and we can invalidate the quicklaunch jar files by changing it. For version 0.10.0 let's deal with getting the API hashed out and a UI rewrite (which is getting more and more needed) can be done for version 0.10.x.

**Comment by WalterLPearson** [comment]
OK.  #4 will be minimalist with GUI needs, I will keep notes on how the GUI could/should change.

You can open a new feature request for the GUI.  I tend to stick to purely functional GUI design (a career in manufacturing, warehouse management, medical, and military software... no glitz, no glamour, minimal form, all function).

Unrelated, please also open a feature request for adding merge support for the missing XML files, so I have something trivial on a separate branch to pop over to when getting burned out.
I am also tinkering with how to handle Tech Tree merging, which will quickly become a problem when we add support for adding Tech and Trees.

**Comment by Tahvohck** [comment]
Functional is good. Fanciness can come later.

Regarding the feature requests - no need for those to be explicit. I'd say go ahead and do whatever tinkering you need to keep engaged, just make sure to switch branches as needed. Technically I'm not a fan of supporting the old merge method anyway since it's destructive and XPath doesn't need special code from our end.

With that said, hooking in the other xml files into the XPath method could probably be done, but I need to have a better understanding of `merge.py` before adding those. Ostensibly it _should_ be as easy as adding them to the `loader.assets.libray.py::PATCHABLE_XML_FILES` array.

**Comment by WalterLPearson** [comment]
I agree that merge-versus-patch needs to be rethought.
If we drop the old merge method, how will _new_ content be added?

If we use the "Add" in xpath, then mod developers will have a hard time putting duplicated content in multiple mod packs.
Maybe we will need a "AddIfMissing" or something?

**Comment by WalterLPearson** [comment]
Finished up basic support for user config and pushed it to my branch.
Moving on to basic GUI.

**Comment by Tahvohck** [comment]
> I agree that merge-versus-patch needs to be rethought.
> If we drop the old merge method, how will _new_ content be added?

Dropping the feature outright won't happen any time soon, as it has its uses, but adding features to it is probably out. As for adding via XPath, extending the PatchOperation method to support the full range of operations (including grouping and tests) is the way to go. With that said ideally we'll eventually decouple the modloader data structure from the base game data structure, so modders don't need the base game ID numbers and such. I have thoughts on that but it's out of scope for this issue for now.

**Comment by WalterLPearson** [comment]
Working code and _most_ of the GUI have been committed for a while.
Getting the tkinter GUI to size correctly is my roadblock.
I googled every tkinter source I can find, but setting minimum and maximum sizes appears impossible.
Would you please take a look and tell me how to fix the GUI sizing?

**Comment by WalterLPearson** [comment]
GUI problems fixed.  Example mod "Electric Slide" working.  All committed to branch.

**Comment by WalterLPearson** [comment]
This is complete enough to deploy.
Variables have no validation, and data types are not implemented, but everything works.
Two example implementations exist.  "Electric Slide" and "Engine Tuner".  I may add more for other common tweaks.

**Comment by WalterLPearson** [comment]
Implemented in develop
